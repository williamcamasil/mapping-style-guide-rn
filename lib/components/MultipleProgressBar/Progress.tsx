import React, { useMemo } from 'react';
import { StyleProp, ViewStyle } from 'react-native';

import { LinearGradient } from 'expo-linear-gradient';

import { useStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

type ProgressPropsType = {
  progress: number;
  theme: AppThemeType;
  startColor: ColorPaletesType;
  endColor: ColorPaletesType;
  style?: StyleProp<ViewStyle>;
};

const startLinearGradient = { x: 0, y: 0 };
const endLinearGradient = { x: 1, y: 0 };

const Progress: React.FC<ProgressPropsType> = ({
  progress, theme, startColor, endColor, style,
}) => {
  const colorGradient = useMemo(
    () => [theme.colors[startColor], theme.colors[endColor]],
    [endColor, startColor, theme],
  );

  const styleBar = useStyles(
    () => [
      {
        flex: progress,
        height: theme.spacings.sNano,
        borderRadius: theme.spacings.sNano / 2,
      },
      style,
    ],
    [progress, style, theme.spacings.sNano],
  );

  return (
    <LinearGradient
      testID="progress-linear-gradient"
      colors={colorGradient}
      start={startLinearGradient}
      end={endLinearGradient}
      style={styleBar}
    />
  );
};

Progress.defaultProps = {
  progress: 0,
};

export default withTheme(Progress);
