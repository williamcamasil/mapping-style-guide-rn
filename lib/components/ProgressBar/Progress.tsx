import React, { useMemo } from 'react';

import { LinearGradient } from 'expo-linear-gradient';

import { useStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

type ProgressPropsType = {
  progress: number;
  theme: AppThemeType;
  startColor?: ColorPaletesType;
  endColor?: ColorPaletesType;
};

const startLinearGradient = { x: 0, y: 0 };
const endLinearGradient = { x: 1, y: 0 };

const Progress: React.FC<ProgressPropsType> = ({
  progress, theme, startColor, endColor,
}) => {
  const colorGradient = useMemo(
    () => [theme.colors[startColor ?? 'primary600'], theme.colors[endColor ?? 'primaryMain']],
    [endColor, startColor, theme],
  );

  const styleBar = useStyles(
    () => [
      {
        width: progress >= 0 && progress <= 100 ? `${progress}%` : 0,
        height: '100%',
        borderRadius: 50,
      },
    ],
    [progress],
  );

  return (
    <LinearGradient
      testID="progress-linear-gradient"
      colors={colorGradient}
      start={startLinearGradient}
      end={endLinearGradient}
      style={styleBar}
    />
  );
};

Progress.defaultProps = {
  progress: 0,
};

export default withTheme(Progress);
