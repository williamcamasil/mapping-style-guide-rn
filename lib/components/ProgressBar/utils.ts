import { ViewStyle } from 'react-native';

export type ProgressSizeType = 'medium' | 'small';

export const getSizeStyle = (size?: ProgressSizeType): ViewStyle => {
  switch (size) {
    case 'small':
      return {
        height: 4,
        borderWidth: 1,
        borderRadius: 2,
      };
    case 'medium':
    default:
      return {
        height: 12,
        borderWidth: 2,
        borderRadius: 6,
      };
  }
};
