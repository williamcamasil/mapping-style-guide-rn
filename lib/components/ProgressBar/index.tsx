import React from 'react';
import {
  StyleProp, StyleSheet, View, ViewStyle,
} from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';
import Progress from './Progress';
import { getSizeStyle, ProgressSizeType } from './utils';

export type ProgressBarType = {
  theme: AppThemeType;
  progress: number;
  size?: ProgressSizeType;
  style?: StyleProp<ViewStyle>;
  backgroundColor?: ColorPaletesType;
  startColor?: ColorPaletesType;
  endColor?: ColorPaletesType;
};

const styled = StyleSheet.create({
  container: {
    justifyContent: 'center',
    overflow: 'hidden',
  },
});

const ProgressBar: React.FC<ProgressBarType> = ({
  theme, progress, size, style, backgroundColor, startColor, endColor,
}) => {
  const styledContainer = useViewStyles(
    () => [
      styled.container,
      {
        borderColor: backgroundColor ?? theme.colors.neutralGray100,
        backgroundColor: backgroundColor ?? theme.colors.neutralGray100,
      },
      getSizeStyle(size),
      style,
    ],
    [backgroundColor, theme, style, size],
  );

  return (
    <View style={styledContainer} testID="progress-bar-container">
      <Progress progress={progress} startColor={startColor} endColor={endColor} />
    </View>
  );
};

ProgressBar.defaultProps = {
  size: 'medium',
};

export default withTheme(ProgressBar);
