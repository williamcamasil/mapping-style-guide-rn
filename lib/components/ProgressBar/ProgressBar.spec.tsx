import React from 'react';
import { StyleSheet } from 'react-native';

import { render, within } from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import ProgressBar from '.';
import { DefaultTheme } from '../../theme';
import Progress from './Progress';

describe('ProgressBar events', () => {
  it('Should verify render the component', () => {
    const result = renderer.create(<ProgressBar progress={0} theme={DefaultTheme} size="medium" />);

    expect(result).toBeDefined();
  });

  it('Should input values to with value 0% in Progress', () => {
    const result = render(<ProgressBar progress={0} size="medium" />);

    const progressBarContainer = result.getByTestId('progress-bar-container');
    const progressBar = within(progressBarContainer).UNSAFE_getByType(Progress);

    expect(progressBar.props).toHaveProperty('progress', 0);
  });

  it('Should input values to with value 50% in Progress', () => {
    const result = render(<ProgressBar progress={50} size="medium" />);

    const progressBarContainer = result.getByTestId('progress-bar-container');
    const progressBar = within(progressBarContainer).UNSAFE_getByType(Progress);

    expect(progressBar.props).toHaveProperty('progress', 50);
  });

  it('Should input values to with value 100% in Progress', () => {
    const result = render(<ProgressBar progress={100} size="medium" />);

    const progressBarContainer = result.getByTestId('progress-bar-container');
    const progressBar = within(progressBarContainer).UNSAFE_getByType(Progress);

    expect(progressBar.props).toHaveProperty('progress', 100);
  });

  it('Should have correct width', () => {
    const result = render(<ProgressBar progress={50} size="medium" />);

    const linearGradient = result.getByTestId('progress-linear-gradient');

    const linearGradientStyles = StyleSheet.flatten(linearGradient.props.style);

    expect(linearGradientStyles.width).toBe('50%');
  });

  it('Should ignore negative numbers', () => {
    const result = render(<ProgressBar progress={-10} size="medium" />);

    const linearGradient = result.getByTestId('progress-linear-gradient');

    const linearGradientStyles = StyleSheet.flatten(linearGradient.props.style);

    expect(linearGradientStyles.width).toBe(0);
  });

  it('Should ignore excessive numbers', () => {
    const result = render(<ProgressBar progress={110} size="medium" />);

    const linearGradient = result.getByTestId('progress-linear-gradient');

    const linearGradientStyles = StyleSheet.flatten(linearGradient.props.style);

    expect(linearGradientStyles.width).toBe(0);
  });

  it('Should change type for medium and verify props', () => {
    const result = render(<ProgressBar progress={10} size="medium" />);

    const progress = result.getByTestId('progress-bar-container');

    const progressStyles = StyleSheet.flatten(progress.props.style);

    expect(progressStyles).toHaveProperty('height', 12);
    expect(progressStyles).toHaveProperty('borderWidth', 2);
    expect(progressStyles).toHaveProperty('borderRadius', 6);
  });

  it('Should change type for small and verify props', () => {
    const result = render(<ProgressBar progress={10} size="small" />);

    const progress = result.getByTestId('progress-bar-container');

    const progressStyles = StyleSheet.flatten(progress.props.style);

    expect(progressStyles).toHaveProperty('height', 4);
    expect(progressStyles).toHaveProperty('borderWidth', 1);
    expect(progressStyles).toHaveProperty('borderRadius', 2);
  });
});

describe('ProgressBar snapshots', () => {
  it('default', () => {
    const tree = renderer.create(<ProgressBar progress={0} size="medium" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('50% in progress', () => {
    const tree = renderer.create(<ProgressBar progress={50} size="medium" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('100% in progress', () => {
    const tree = renderer.create(<ProgressBar progress={100} size="medium" />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
