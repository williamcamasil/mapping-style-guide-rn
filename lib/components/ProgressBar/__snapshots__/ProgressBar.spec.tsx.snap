// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar snapshots 50% in progress 1`] = `
<View
  style={
    {
      "backgroundColor": "#F7F7F8",
      "borderColor": "#F7F7F8",
      "borderRadius": 6,
      "borderWidth": 2,
      "height": 12,
      "justifyContent": "center",
      "overflow": "hidden",
    }
  }
  testID="progress-bar-container"
>
  <View
    colors={
      [
        "#2453A9",
        "#013188",
      ]
    }
    end={
      {
        "x": 1,
        "y": 0,
      }
    }
    start={
      {
        "x": 0,
        "y": 0,
      }
    }
    style={
      {
        "borderRadius": 50,
        "height": "100%",
        "width": "50%",
      }
    }
    testID="progress-linear-gradient"
  />
</View>
`;

exports[`ProgressBar snapshots 100% in progress 1`] = `
<View
  style={
    {
      "backgroundColor": "#F7F7F8",
      "borderColor": "#F7F7F8",
      "borderRadius": 6,
      "borderWidth": 2,
      "height": 12,
      "justifyContent": "center",
      "overflow": "hidden",
    }
  }
  testID="progress-bar-container"
>
  <View
    colors={
      [
        "#2453A9",
        "#013188",
      ]
    }
    end={
      {
        "x": 1,
        "y": 0,
      }
    }
    start={
      {
        "x": 0,
        "y": 0,
      }
    }
    style={
      {
        "borderRadius": 50,
        "height": "100%",
        "width": "100%",
      }
    }
    testID="progress-linear-gradient"
  />
</View>
`;

exports[`ProgressBar snapshots default 1`] = `
<View
  style={
    {
      "backgroundColor": "#F7F7F8",
      "borderColor": "#F7F7F8",
      "borderRadius": 6,
      "borderWidth": 2,
      "height": 12,
      "justifyContent": "center",
      "overflow": "hidden",
    }
  }
  testID="progress-bar-container"
>
  <View
    colors={
      [
        "#2453A9",
        "#013188",
      ]
    }
    end={
      {
        "x": 1,
        "y": 0,
      }
    }
    start={
      {
        "x": 0,
        "y": 0,
      }
    }
    style={
      {
        "borderRadius": 50,
        "height": "100%",
        "width": "0%",
      }
    }
    testID="progress-linear-gradient"
  />
</View>
`;
